src/action/base/create_or_merge_nix_config.rs

```diff
@@ -0,0 +1,490 @@
+    ParseNixConfig(#[from] nix_config_parser::ParseError),
+    #[error("Could not merge Nix configuration for key(s) {}; consider removing them from {1} in your editor, or removing your existing configuration with `rm {1}`",
+        .0
+        .iter()
+        .map(|v| format!("`{v}`"))
+        .collect::<Vec<_>>()
+        .join(", "))]
+    UnmergeableConfig(Vec<String>, std::path::PathBuf),
+}
+
+/// Create or merge an existing `nix.conf` at the specified path.
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub struct CreateOrMergeNixConfig {
+    pub(crate) path: PathBuf,
+    nix_configs: NixConfigs,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+struct NixConfigs {
+    existing_nix_config: Option<NixConfig>,
+    merged_nix_config: NixConfig,
+}
+
+impl CreateOrMergeNixConfig {
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn plan(
+        path: impl AsRef<Path>,
+        pending_nix_config: NixConfig,
+    ) -> Result<StatefulAction<Self>, ActionError> {
+        let path = path.as_ref().to_path_buf();
+
+        let nix_configs = NixConfigs {
+            existing_nix_config: None,
+            merged_nix_config: NixConfig::new(),
+        };
+
+        let mut this = Self { path, nix_configs };
+
+        if this.path.exists() {
+            let metadata = this
+                .path
+                .metadata()
+                .map_err(|e| ActionError::GettingMetadata(this.path.clone(), e))?;
+
+            if !metadata.is_file() {
+                return Err(ActionError::PathWasNotFile(this.path));
+            }
+
+            // Does the file have the right permissions?
+            let discovered_mode = metadata.permissions().mode();
+            // We only care about user-group-other permissions
+            let discovered_mode = discovered_mode & 0o777;
+
+            if discovered_mode != NIX_CONF_MODE {
+                return Err(ActionError::PathModeMismatch(
+                    this.path,
+                    discovered_mode,
+                    NIX_CONF_MODE,
+                ));
+            }
+
+            let existing_nix_config = NixConfig::parse_file(&this.path)
+                .map_err(CreateOrMergeNixConfigError::ParseNixConfig)
+                .map_err(|e| ActionError::Custom(Box::new(e)))?;
+            this.nix_configs.existing_nix_config = Some(existing_nix_config.clone());
+            let mut merged_nix_config = NixConfig::new();
+            let mut unmergeable_config_names = Vec::new();
+
+            for (pending_conf_name, pending_conf_value) in pending_nix_config.settings() {
+                if let Some(existing_conf_value) =
+                    existing_nix_config.settings().get(pending_conf_name)
+                {
+                    let pending_conf_value = pending_conf_value.split(' ').collect::<Vec<_>>();
+                    let existing_conf_value = existing_conf_value.split(' ').collect::<Vec<_>>();
```

Comment:
I am thinking of `nix-path` and `extra-nix-path` here, what if they have paths with spaces in them? `/temp\ dir/nix`

Reply:
Nix doesn't support this, so I think it's fine that that would produce a bad value?

```
$ pwd
/home/vin/workspace/vcs/nixpkgs/test nixpkgs
$ nix build --impure --expr '(import <nixpkgs> {})' --nix-path nixpkgs="$PWD"
warning: Nix search path entry '/home/vin/workspace/vcs/nixpkgs/test' does not exist, ignoring
warning: Nix search path entry 'nixpkgs' does not exist, ignoring
error:
       … while calling the 'import' builtin

         at «string»:1:2:

            1| (import <nixpkgs> {})
             |  ^

       … while realising the context of a path

         at «none»:0: (source not available)

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

       at «none»:0: (source not available)
```

Reply:
Ah, ok!